### ISSUE TITLE
Test failure when building against HDF5 1.8.16 on Python 2.7.11

### ISSUE THREAD

**Issue by jakirkham** [h5py/h5py#696]
Seeing the following test failure in 2.6.0 when building against HDF5 1.8.16 on Python 2.7.11. This happens in a [docker container](https://github.com/pelson/Obvious-CI/blob/master/obvious-ci.docker/linux64_obvci/Dockerfile) (CentOS 6). For more details about when this is occurring see this [log](https://circleci.com/gh/conda-forge/staged-recipes/901). 

```
======================================================================
FAIL: test_default (h5py.tests.old.test_file.TestFileOpen)
Default semantics in the presence or absence of a file
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/envs/_test/lib/python2.7/site-packages/h5py/tests/old/test_file.py", line 48, in test_default
    self.assertEqual(f.mode, 'r')
AssertionError: 'r+' != 'r'
```

Related: https://github.com/h5py/h5py/issues/554
Related: https://github.com/conda-forge/staged-recipes/pull/201

**Comment by jreadey** [comment]
I'm not seeing this with a Docker container consisting of HDF5 1.8.16, h5py 2.6.0, and Python 3.5.1.

Is this the scenario:

```
import h5py
f = h5py.File("afile.h5", 'r')
print(f.mode)   # returns 'r' for me
```
?

I don't have an image with Python 2.7.1 handy.

**Comment by andrewcollette** [comment]
@jreadey, I think the problem is more what happens when no explicit mode is used; for example:
```
f = h5py.File("afile.h5")
```
If `afile.h5` has readonly permissions, `f.mode` is supposed to be `r`, but for some reason it's not.

**Comment by jreadey** [comment]
@andrewcollette, right I noticed that if I open a read-only file as root (in docker processes are often run as root), the mode will come back as r+.  I made a fix in h5serv for this just last week: https://github.com/HDFGroup/h5serv/commit/ab3bd7b5e1bace2185d4e14470015880a5d3e06b.  For the issue here, I suspect the change would need to be made in the hdf5 library itself.  (or change to test to run under a non-root user)

**Comment by jakirkham** [comment]
Given @jreadey's [comment]( https://github.com/h5py/h5py/issues/696#issuecomment-202500480 ), would a reasonable workaround just be to set the option to `r` explicitly if not selected?

**Comment by matthew-brett** [comment]
I am getting the same test error for manylinux1 wheels, h5py 2.6.0 - again in a docker container.

I wonder whether it be worth adapting the test to look for root permissions before doing the check?

**Comment by tacaswell** [comment]
Closed by #724
